dso-l2/src/main/java/com/tc/objectserver/entity/ActivePassiveAckWaiter.java

```diff
@@ -60,6 +61,25 @@ public synchronized void waitForReceived() {
       throw new RuntimeException(ie);
     }
   }
+  
+  public void runWhenCompleted(Runnable r) {
+    boolean wasSet = false;
+    synchronized (this) {
+      if (!this.completedPending.isEmpty()) {
+        finalizer = r;
+        wasSet = true;
+      }
+    }
+    if (!wasSet) {
+      r.run();
+    }
+  }
+  
+  public void run() {
+    if (finalizer != null) {
```

Comment:
no visibility issues?

Reply:
No.  Implemented this way to keep the run out of they synchronization block but I will move the execution in this class so that it is more clear when the finalizer is executed.

Reply:
added a comment about visibility and updated the code